src/install/install.zig

```diff
@@ -10093,162 +10103,452 @@ pub const PackageManager = struct {
-                    if (version) |v| {
-                        const label = std.fmt.bufPrint(buf[0..], "{}", .{pkg.resolution.fmt(strbuf, .posix)}) catch @panic("Resolution name too long");
-                        if (std.mem.eql(u8, label, v)) break :brk pkg_id;
-                    } else {
-                        first_match = pkg_id;
-                    }
-                } else break;
-            }
-            if (first_match) |id| {
-                if (matches_count > 1) {
-                    Output.prettyErrorln(
-                        "\n<r><red>error<r>: please specify a precise version:<r>",
-                        .{},
-                    );
-                    i = 0;
-                    while (i < manager.lockfile.packages.len) {
-                        if (std.mem.indexOfScalar(u64, pkg_hashes[i..], name_hash)) |idx| {
-                            defer i += idx + 1;
-                            const pkg_id = i + idx;
-                            const pkg = manager.lockfile.packages.get(pkg_id);
-                            if (!std.mem.eql(u8, pkg.name.slice(strbuf), name)) continue;
+            const dependencies = lockfile.buffers.dependencies.items;
+
+            var matches_found: u32 = 0;
+            var maybe_first_match: ?struct { DependencyID, PackageID } = null;
+            for (dependencies, 0..) |dep, dep_id| {
+                if (dep.name_hash != name_hash) continue;
+                matches_found += 1;
+                const pkg_id = lockfile.buffers.resolutions.items[dep_id];
+                if (pkg_id == invalid_package_id) continue;
+                const pkg = lockfile.packages.get(pkg_id);
+                if (version) |v| {
+                    const label = std.fmt.bufPrint(buf[0..], "{}", .{pkg.resolution.fmt(strbuf, .posix)}) catch @panic("Resolution name too long");
+                    if (std.mem.eql(u8, label, v)) break :brk .{ @intCast(dep_id), pkg_id };
+                } else maybe_first_match = .{ @intCast(dep_id), pkg_id };
+            }
+
+            const first_match = maybe_first_match orelse {
+                Output.prettyErrorln("\n<r><red>error<r>: package <b>{s}<r> not found<r>", .{pkg_maybe_version_to_patch});
+                Global.crash();
+                return;
+            };
 
-                            Output.prettyError("  {s}@<blue>{}<r>\n", .{ pkg.name.slice(strbuf), pkg.resolution.fmt(strbuf, .posix) });
-                        } else break;
-                    }
-                    Output.flush();
-                    Global.crash();
-                    return;
+            if (matches_found > 1) {
+                Output.prettyErrorln(
+                    "\n<r><red>error<r>: Found multiple versions of <b>{s}<r>, please specify a precise version from the following list:<r>\n",
+                    .{name},
+                );
+                var i: usize = 0;
+                const pkg_hashes = lockfile.packages.items(.name_hash);
```

Comment:
in the future, this is where you would get the `package_id` for the current workspace and only loop through it's packages, instead of all packages